fix: set a failed canceled job status correctly

Kira-Pilot opened the pull request:
resolves #1374

When a previous job was canceled and that cancellation was unsuccessful, we incorrectly set the job status as `canceled`.  This fix remedies that issue by setting the status as `Failed`.

Kira-Pilot committed:
set a failed canceled job status correctly

resolves #1374

Kira-Pilot committed:
set a failed canceled job status correctly

resolves #1374

mafredri committed:
fix: Increase randomness for names used in tests (#3063)

We are starting to run into test flakes due to lack of randomness in CI,
this change simply bumps randomness by additional suffix numbers.

See: https://github.com/coder/coder/issues/3038#issuecomment-1190283608

mafredri committed:
chore: Speed up port-forward tests (#3062)

* chore: Speed up port-forward tests

* chore: Add t.Helper and ensure listener closure on error

Kira-Pilot committed:
clarify start validation in schedule (#3052)

resolves #2792

dependabot[bot] committed:
chore: bump github.com/spf13/afero from 1.9.0 to 1.9.2 (#3046)

Bumps [github.com/spf13/afero](https://github.com/spf13/afero) from 1.9.0 to 1.9.2.
- [Release notes](https://github.com/spf13/afero/releases)
- [Commits](https://github.com/spf13/afero/compare/v1.9.0...v1.9.2)

---
updated-dependencies:
- dependency-name: github.com/spf13/afero
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

mafredri committed:
fix: Improve TestSSH reliability on macOS (#3067)

Related issue: https://github.com/coder/coder/issues/2122

ammario committed:
ci: skip long jobs when only docs change (#3072)

dependabot[bot] committed:
chore: bump terser from 4.8.0 to 4.8.1 in /site (#3068)

Bumps [terser](https://github.com/terser/terser) from 4.8.0 to 4.8.1.
- [Release notes](https://github.com/terser/terser/releases)
- [Changelog](https://github.com/terser/terser/blob/master/CHANGELOG.md)
- [Commits](https://github.com/terser/terser/commits)

---
updated-dependencies:
- dependency-name: terser
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Commit:
fix: Statuses breaking line in the UI (#3071)

* fix: Fix statuses breaking line in the UI

* fix: AppLink stories

ammario committed:
ci: fix stale issue workflow (#3073)

mafredri committed:
fix: Increase test timeout for TestCreate/CreateFromListWithSkip (#3077)

Considering database load and CI performance during testing, we should
avoid failing too early.

mafredri committed:
fix: Add `continue-on-error` to codecov action step (#3081)

Avoid relying on codecov to manage action step failure, hopefully works
around:
https://github.com/codecov/codecov-action/issues/788

mafredri committed:
fix: Increase CI timeout for postgres test (#3079)

The Go test timeout uses 20m, if we want to get a stack trace, we must
allow the actions worker to run longer than that.

sreya committed:
feat: add verbose error messaging (#3053)

Kira-Pilot committed:
added unit test for convertProvisionerJob

Kira-Pilot committed:
added unit test for convertProvisionerJob

Kira-Pilot commented on coderd/templateversions_test.go:
```diff
@@ -151,7 +157,7 @@ func TestPatchCancelTemplateVersion(t *testing.T) {
 			version, err = client.TemplateVersion(context.Background(), version.ID)
 			require.NoError(t, err)
-			return version.Job.Status == codersdk.ProvisionerJobCanceled
+			return version.Job.Status == codersdk.ProvisionerJobFailed // cancel times out
```
Does anyone know why this might be timing out? I don't think this is related to the changes I made (but could be wrong). 
![Screen Shot 2022-07-20 at 10 32 35 AM](https://user-images.githubusercontent.com/19142439/180079203-ceef1f0a-1e74-4c4c-96f4-ae7b09399f5e.png)

johnstcn replied:
I checked on main (with some changes to remove that pesky `require.Eventually`) and it looks like this test is already subtly broken:

![image](https://user-images.githubusercontent.com/4949514/180197017-1fe3e229-eafc-424b-876f-d6c8ace2a1d5.png)

I've filed a separate issue https://github.com/coder/coder/issues/3083 for this; you can leave this as-is for now and I'll follow up on that.

Kira-Pilot commented on coderd/workspacebuilds_test.go:
```diff
@@ -229,7 +229,7 @@ func TestPatchCancelWorkspaceBuild(t *testing.T) {
 		build, err = client.WorkspaceBuild(context.Background(), build.ID)
 		require.NoError(t, err)
-		return build.Job.Status == codersdk.ProvisionerJobCanceled
+		return build.Job.Status == codersdk.ProvisionerJobFailed // this is timing out
```
Ditto

johnstcn replied:
Probably the same reason as above! I'll take a look in https://github.com/coder/coder/issues/3083

Kira-Pilot committed:
Update coderd/provisionerjobs_internal_test.go

Co-authored-by: Cian Johnston <cian@coder.com>

Kira-Pilot committed:
Update coderd/provisionerjobs_internal_test.go

Co-authored-by: Cian Johnston <cian@coder.com>

mafredri committed:
fix: Deadlock and race in `peer`, test improvements (#3086)

* fix: Potential deadlock in peer.Channel dc.OnOpen

* fix: Potential send on closed channel

* fix: Improve robustness of waitOpened during close

* chore: Simplify statements

* fix: Improve teardown and timeout of peer tests

* fix: Improve robustness of TestConn/Buffering test

* Update peer/channel.go

Co-authored-by: Steven Masley <Emyrk@users.noreply.github.com>

Commit:
refactor: Initial color palette changes (#3087)

sreya committed:
fix: avoid emitting version warning when connection error encountered (#3082)

johnstcn committed:
fix: coderdtest: increase ForceCancelInterval (#3085)

Two coderd unit tests (TestPatchCancelTemplateVersion/Success and TestPatchCancelWorkspaceBuild) implied erroneously that the job was canceled successfully.

This is not the case, as these unit tests do not include a Provision_Complete response in the input to the
echo provisioner. Now explicitly checking the job error and bumping the force cancel interval to be longer.

Fixes #3083.

johnstcn commented on coderd/templateversions_test.go:
```diff
@@ -125,6 +125,12 @@ func TestPatchCancelTemplateVersion(t *testing.T) {
+		require.Eventually(t, func() bool {
+			var err error
+			version, err = client.TemplateVersion(context.Background(), version.ID)
+			require.NoError(t, err)
```
This is a thing that I like to keep a watch out for as it burned me a couple times in the past with flaky tests -- `require.Eventually` runs its callback in a separate goroutine to the testmain goroutine. Any function from `github.com/stretchr/testify/require` can end up calling `t.FailNow` which is unsafe to do in anything besides the testmain. However each `require` has a corresponding `assert` that doesn't have this problem.

https://pkg.go.dev/testing#T.FailNow

As a bonus, the `assert.*` versions can also be chained as booleans, which lets you write neat stuff like this:

```suggestion
			return assert.NoError(t, err) &&
			    version.Job.Status == codersdk.ProvisionerJobFailed
```

kylecarbs replied:
I can attest to this being sucky! I believe @johnstcn fixed up all the times I messed this up, and I didn't know about it until he pointed it out! 🙈

Kira-Pilot replied:
Ahh, good to know. Thanks you for the explanation and snippet.

mafredri replied:
Seconded (or is that thirded)! I’ve also noticed there are other dangers to using `require` because of it’s use of `FailNow`. For example, it’s easy to miss that `require` is essentially `if err != nil { return }`, so assumptions like a connection will be closed further down the line break, and it diverges from the mental model of regular Go code.

Kira-Pilot committed:
conflicts

Kira-Pilot committed:
Update coderd/templateversions_test.go

Co-authored-by: Cian Johnston <cian@coder.com>

Kira-Pilot commented:
Wow, what a rebase effort. Closing in favor of a cleaner #3101.
